index/scorch/snapshot_index.go

```diff
@@ -404,11 +404,42 @@ func (i *IndexSnapshot) DocumentVisitFieldTerms(id index.IndexInternalID,
 	ss := i.segment[segmentIndex]
 
 	if zaps, ok := ss.segment.(segment.DocumentFieldTermVisitable); ok {
-		return zaps.VisitDocumentFieldTerms(localDocNum, fields, visitor)
+		// get the list of doc value persisted fields
+		pFields, err := zaps.VisitableDocValueFields()
+		if err != nil {
+			return err
+		}
+		// assort the fields for which terms look up have to
+		// be performed runtime
+		fieldsDvPending := stringsRemoveStrings(fields, pFields)
+		if len(fieldsDvPending) == 0 {
+			// all fields are doc value persisted
+			return zaps.VisitDocumentFieldTerms(localDocNum, fields, visitor)
+		}
+
+		// concurrently trigger the runtime doc value computations for pending fields
+		// as well as the visit of the persisted doc values
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go visitDocumentFieldCacheAsync(localDocNum, fieldsDvPending, ss, &wg, visitor)
```

Comment:
ha...may be not a good idea to call the visitor callback from different routines. Looks like the FieldCache preparation can be done concurrently and the invoking the visitor callback can be in sequential..

Reply:
another problem here is that you ignore errors that happen because it's async and you have no easy way to deal with them.  this is usually a bad sign, and done right, you'd need a channel to return the errors on.

Reply:
reading this more carefully, yes, we must only invoke the callback from a single goroutine.  the reason is that the callbacks are written in a simple way expecting this, if multiple goroutines call them, they would need locking/synchronization on all of their data.